File: php/src/Application/Command/MoveRover/MoveRover.php
<?php
declare(strict_types=1);

namespace App\Application\Command\MoveRover;

use App\Infrastructure\Storage;

class MoveRover
{
    /** @var Storage */
    private $storage;
    /** @var int */
    private $maxX;
    /** @var int */
    private $maxY;

    public function __construct(Storage $storage)
    {
        $this->storage = $storage;
    }

    public function move(string $roverId, string $movements)
    {
        $roverData = $this->storage->read($roverId);

        if ($roverData !== null) {

            $plateau = $this->storage->read($roverData['plateau-id']);
            if ($plateau !== null) {

                $position           = $roverData['position'];
                $positionArray      = explode(',', $position);
                $initialX           = $positionArray[0];
                $initialY           = $positionArray[1];
                $initialOrientation = $positionArray[2];

                $movements = str_split($movements);

                $currentX           = $initialX;
                $currentY           = $initialY;
                $currentOrientation = $initialOrientation;

                $plateauDimensions      = $plateau['dimensions'];
                $plateauDimensionsArray = explode(',', $plateau['dimensions']);
                $this->maxX             = $plateauDimensionsArray[0];
                $this->maxY             = $plateauDimensionsArray[1];

                foreach ($movements as $movement) {
                    switch ($movement) {
                        case 'R':
                            if ($currentOrientation === 'N') {
                                $currentOrientation = 'E';
                            }
                            break;

                        case 'M':
                            if ($currentOrientation === 'E') {
                                $currentX++;
                            }
                            break;
                    }

                    $roverData['position'] = implode(',', [$currentX, $currentY, $currentOrientation]);

                    $this->storage->store(
                        $roverId,
                        $roverData
                    );
                }
            } else {
                throw new \Exception('Plateau not found');
            }
        } else {
            throw new \Exception('Rover not found');
        }
    }
}

File: csharp/MarsRoverKata/Application/Command/MoveRover.cs
using MarsRoverKata.Application.Infrastructure;
using System;
using System.Linq;

namespace MarsRoverKata.Application.Command
{
    public class MoveRover
    {
        private GenericStorage _storage;
        private string maxX;
        private string maxY;

        public MoveRover(GenericStorage storage)
        {
            _storage = storage;
        }

        public void Move(string roverId, string movements)
        {
            object roverObj = _storage.Read(roverId);
            
            if (roverObj != null)
            {
                var roverData = (Rover)roverObj;
                var plateauObj = _storage.Read(roverData.PlateauId);

                if (plateauObj != null)
                {
                    var plateau = (Plateau)plateauObj;

                    var position = roverData.Position;
                    var positionArray = position.Split(',');
                    var initialX = int.Parse(positionArray[0]);
                    var initialY = int.Parse(positionArray[1]);
                    var initialOrientation = positionArray[2];

                    var movementsArray = movements.ToArray();

                    var currentX = initialX;
                    var currentY = initialY;
                    var currentOrientation = initialOrientation;
                    var plateauDimensions = plateau.Dimensions;
                    var plateauDimensionsArray = plateau.Dimensions.Split(',');
                    maxX = plateauDimensionsArray[0];
                    maxY = plateauDimensionsArray[1];

                    foreach (var movement in movementsArray) {
                        switch (movement) {
                            case 'R':
                                if (currentOrientation == "N") {
                                    currentOrientation = "E";
                                }
                                break;
                            case 'M':
                                if (currentOrientation == "E") {
                                    currentX++;
                                }
                                break;
                        }

                        roverData.Position = string.Join(',', new string[] { currentX.ToString(), currentY.ToString(), currentOrientation });
                    }
                }
                else
                {
                    throw new Exception("Plateau not found");
                }
            }
            else
            {
                throw new Exception("Rover not found");
            }
        }
    }

    public class Plateau
    {
        public string Dimensions { get; set; }
    }

    public class Rover
    {
        public string Position { get; set; }
        public string PlateauId { get; set; }
    }
}
